### ISSUE TITLE
Problem detecting libc

### ISSUE THREAD

**Issue by repomaa** [crystal-lang/crystal#7946]
I noticed a regression in iconv when built with musl. This is actually caused by crystal not being able to detect the libc when building and then defaults to gnu. This in turn is caused by a regression of ldd in musl which doesn't recognize the `--version` flag anymore and fails. See here: https://github.com/crystal-lang/crystal/blob/b038213e636c055ad7fdf98d0fc294200ab849a5/src/compiler/crystal/config.cr#L60-L63 how crystal handles this.

This raises the question why crystal isn't using a build flag to bake the libc variant into the compiler as a constant when it's built for the platform. It's not possible to use the crystal compiler linked against gnu libc to build for musl and vice versa, is it?

See the discussion with alpine maintainers at https://dev.alpinelinux.org/irclogs/%23alpine-linux-2019-07.log from `2019-07-02 13:32:11` on.

**Comment by ysbaddaden** [comment]
We want to detect at runtime whether the binary wants to build against musl-libc or glibc, because we build a static binary for Crystal releases on Alpine Linux, then use it on whatever Linux distribution.

It hack should have been safe, because we follow what `autoconf` or `automake` does to detect musl, but apparently this has changed :/

Please look at the Git history to find the merge of the change and the actual discussion to understand more.

**Comment by repomaa** [comment]
For now a workaround to be able to build for musl is to add `-Dmusl` to the `crystal build` or `shards` commands.

**Comment by bcardiff** [comment]
From

```
2019-07-02 04:44:22 <mikeee__> running 'ldd --version' results in '/lib/ld-musl-x86_64.so.1: cannot load --version: No such file or directory'
2019-07-02 04:44:51 <mikeee__> after https://github.com/alpinelinux/aports/commit/d4a7955cc72b9395779c20755df88f983763c43b#diff-1f663dd56ac26d76d510b53ab0a6c975
2019-07-02 04:45:38 <mikeee__> the problem is that the (now) wrapper script /usr/bin/ldd will turn it into '/lib/ld-xxx --list -- --version'
```

A solution would seems to be to extract the first part of the output of `ldd --version` or check if the string includes (and [not just start](https://github.com/crystal-lang/crystal/blob/b038213e636c055ad7fdf98d0fc294200ab849a5/src/compiler/crystal/config.cr#L66)).

**Comment by straight-shoota** [comment]
> This raises the question why crystal isn't using a build flag to bake the libc variant into the compiler as a constant when it's built for the platform.
> It's not possible to use the crystal compiler linked against gnu libc to build for musl and vice versa, is it?

The libc was hard-coded before #7479, but we wanted more flexibility, Because, in fact the compiler doesn't care whether itself it linked against musl or glibc, it can link to both.

**Comment by straight-shoota** [comment]
The actual issue here is that musl's `ldd` wrapper script is broken. I suppose this is going to be fixed. This behaviour was copied from GNU's [`config.guess`](https://git.savannah.gnu.org/gitweb/?p=config.git;a=blob_plain;f=config.guess), so it's rather commonly used and the breakage is simply on musl. `ldd --version` should just work for both glibc and musl.

If there are any better ideas to tell whether a system uses musl or gnu as default C toolchain, we could certainly work on improving this. But I doubt anyone knows a better and more reliable method.

Apart from that, I don't think there is anything to do right now.

**Comment by straight-shoota** [comment]
This issue has been fixed in alpine 3.11 and we don't need to bother.
